### ISSUE TITLE
Selecting different subset in each iteration

### ISSUE THREAD

**Issue by LalchandPandia** [huggingface/accelerate#3538]
Hi,
I am trying to select a subset of data in each iteration and train a model on this data for 2 epochs. In each iteration, I need to select randomly and ensure that the data is different from previous iteration. 
Is there a cleaner way to do it?
Below is snippet from  my code:
with accelerator.main_process_first():
    train_dataset = train_dataset.map(partial(encode_with_message, tokenizer=tokenizer, max_seq_length=1024, include_response=True,  response_only=False, only_first_two=False, prompt_only=False, add_bos_token=False), batched=False,
                                      remove_columns=[name for name in train_dataset.column_names if name not in ["input_ids", "labels", "attention_mask"]
            ], load_from_cache_file=False, num_proc=args_preprocessing_num_workers,
            desc="Tokenizing and reformatting instruction data",)
    print("After  map:", train_dataset.column_names)
    train_dataset.set_format(type="pt")
    train_dataset = train_dataset.filter(lambda example: (example["labels"] != -100).any())
model, optimizer = accelerator.prepare(model, optimizer)
for i in range(10):
   if accelerator.is_main_process:
        all_indices = list(range(len(train_dataset)))
        selected_indices = random.Random(random_seed + i).sample(all_indices, block_size)
    else:
        selected_indices = None
        
    obj_list = [selected_indices]
    broadcast_object_list(obj_list, src=0)
    selected_indices = obj_list[0]  # now all processes have the same list

    # Subset dataset consistently
    selected_dataset = train_dataset.select(selected_indices)

    if accelerator.is_main_process:
        print(f"Randomly selected {block_size} train instances")
    #randomly choose 50K
    #train_dataset = train_dataset.shuffle(seed=random_seed+i)
    #train_dataset = train_dataset.select(range(block_size))
    #print(f"Randomly selected {block_size} train instances")
    
    train_dataloader = DataLoader(
        selected_dataset,
        shuffle=True,
        collate_fn=DataCollatorForSeq2Seq(tokenizer=tokenizer, model=model, padding="longest"),
        batch_size=per_device_train_batch_size,
    )
   train_dataloader = accelerator.prepare(train_dataloader)
   for epoch in range(starting_epoch, starting_epoch+args_num_train_epochs)
        model.train()
        for step, batch in enumerate(active_dataloader):
           <----loss calculation--->

**Comment by LalchandPandia** [comment]
@muellerzr
